### ISSUE TITLE
[BUG] Storage permission is asked in Android 13 and can't be granted

### ISSUE THREAD

**Issue by WhtUsrWrks** [ankidroid/Anki-Android#14423]
Hi.. I hadn't used ankidroid so my Samsung Galaxy A21 removed its permissions, and I wanted to use it again today but when I open it it says "Grant storage permission", but I've granted all permissions and it still asks. I've tried u installing and reinstalling and it says the same thing. I don't seem to have a dedicated Storage Permission option in permission manager but it has permission for files and that's on... I don't know what else to change? Thank you

**Comment by thedroiddiv** [comment]
@WhtUsrWrks can you share screenshot of the permission screen and the popup where it is asking permissions? Or better maybe attach a screen recording.

**Comment by thedroiddiv** [comment]
Thanks for sharing!

**Comment by BrayanDSO** [comment]
I'm pinning this so we don't forget it

**Comment by Ezkielnet** [comment]
Hello. 
I have the same problem with One Plus Nord 2, Android 13.

**Comment by lukstbit** [comment]
Nobody mentioned which version of the app are they running. Is the app installed from the app store, downloaded from here etc?

**Comment by WhtUsrWrks** [comment]
Mine says 2.16.5 and I downloaded it from Google play

**Comment by Ezkielnet** [comment]
Versionn 2.16.5 and downloaded it from Google play.

**Comment by Ezkielnet** [comment]
https://github.com/ankidroid/Anki-Android/assets/23616105/bd24a159-dedd-4376-a636-93bcefe7d7cc

**Comment by Sc3l3t0n** [comment]
Could this be the problem?
This would at least explain why the system does not ask for memory access in the first place.

> Three new permission is introduced with Android 13: READ_MEDIA_IMAGES, READ_MEDIA_VIDEO, and READ_MEDIA_AUDIO. If you were using READ_EXTERNAL_STORAGE for accessing files, then you need to use one of these new permissions. If a user with Android 13 has previously granted this permission, then all is good. If not, then the permission request for READ_EXTERNAL_STORAGE will be ignored. 

Source: https://medium.com/@muhammadbilal4074/storage-permission-in-android-13-89372e543864#:~:text=Three%20new%20permission%20is%20introduced,permission%2C%20then%20all%20is%20good.

**Comment by queue-tip** [comment]
I experience the same running version 2.16.5 on Pixel 7 Pro with Android 13 (September 2023 update)

**Comment by hmenke** [comment]
Launching AnkiDroid this `logcat` is produced.

```console
09-23 21:21:47.398 18391 19178 I Role    : com.ichi2.anki not qualified for android.app.role.HOME due to missing RequiredComponent{mIntentFilterData=IntentFilterData{mAction='android.intent.action.MAIN', mCategories='[android.intent.category.HOME]', mDataScheme='null', mDataType='null'}, mMinTargetSdkVersion=1, mPermission='null', mQueryFlags=0, mMetaData=[]}
09-23 21:21:47.401 18391 19178 I Role    : com.ichi2.anki not qualified for android.app.role.DIALER due to missing RequiredComponent{mIntentFilterData=IntentFilterData{mAction='android.intent.action.DIAL', mCategories='[]', mDataScheme='null', mDataType='null'}, mMinTargetSdkVersion=1, mPermission='null', mQueryFlags=0, mMetaData=[]}
09-23 21:21:47.404 18391 18642 I Role    : com.ichi2.anki not qualified for android.app.role.SMS due to missing RequiredComponent{mIntentFilterData=IntentFilterData{mAction='android.provider.Telephony.SMS_DELIVER', mCategories='[]', mDataScheme='null', mDataType='null'}, mMinTargetSdkVersion=1, mPermission='android.permission.BROADCAST_SMS', mQueryFlags=0, mMetaData=[]}
09-23 21:21:47.406 20072 20072 E AppInfoDashboard: Package: com.ichi2.anki not found for user: 10
09-23 21:21:47.406 20072 20072 E AppInfoDashboard: Package: com.ichi2.anki not found for user: 11
09-23 21:21:47.407  1494  5310 E AppOps  : Bad call made by uid 1000. Package "com.ichi2.anki" does not belong to uid -1.
09-23 21:21:47.407  1494  5310 E AppOps  : noteOperation
09-23 21:21:47.407  1494  5310 E AppOps  : java.lang.SecurityException: Specified package "com.ichi2.anki" under uid -1 but it is not
09-23 21:21:47.407  1494  5310 E AppOps  : 	at com.android.server.appop.AppOpsService.verifyAndGetBypass(AppOpsService.java:4703)
09-23 21:21:47.407  1494  5310 E AppOps  : 	at com.android.server.appop.AppOpsService.noteOperationUnchecked(AppOpsService.java:3466)
09-23 21:21:47.407  1494  5310 E AppOps  : 	at com.android.server.appop.AppOpsService.noteOperationImpl(AppOpsService.java:3454)
09-23 21:21:47.407  1494  5310 E AppOps  : 	at com.android.server.appop.AppOpsService.-$$Nest$mnoteOperationImpl(Unknown Source:0)
09-23 21:21:47.407  1494  5310 E AppOps  : 	at com.android.server.appop.AppOpsService$CheckOpsDelegateDispatcher.lambda$noteOperation$4(AppOpsService.java:7662)
09-23 21:21:47.407  1494  5310 E AppOps  : 	at com.android.server.appop.AppOpsService$CheckOpsDelegateDispatcher.$r8$lambda$oOwUoG55BBmJ6yiwhp3OAq3tQ30(Unknown Source:0)
09-23 21:21:47.407  1494  5310 E AppOps  : 	at com.android.server.appop.AppOpsService$CheckOpsDelegateDispatcher$$ExternalSyntheticLambda3.apply(Unknown Source:32)
09-23 21:21:47.407  1494  5310 E AppOps  : 	at com.android.server.policy.AppOpsPolicy.noteOperation(AppOpsPolicy.java:225)
09-23 21:21:47.407  1494  5310 E AppOps  : 	at com.android.server.appop.AppOpsService$CheckOpsDelegateDispatcher.noteOperation(AppOpsService.java:7660)
09-23 21:21:47.407  1494  5310 E AppOps  : 	at com.android.server.appop.AppOpsService.noteOperation(AppOpsService.java:3438)
09-23 21:21:47.407  1494  5310 E AppOps  : 	at com.android.internal.app.IAppOpsService$Stub.onTransact(IAppOpsService.java:479)
09-23 21:21:47.407  1494  5310 E AppOps  : 	at android.os.Binder.execTransactInternal(Binder.java:1309)
09-23 21:21:47.407  1494  5310 E AppOps  : 	at android.os.Binder.execTransact(Binder.java:1268)
```

**Comment by dae** [comment]
Maybe related? https://www.reddit.com/r/androiddev/comments/146cd1y/comment/jnt3jmn/

**Comment by gragramel** [comment]
Dear AnkiDroid Team/Users,

I experienced the same with AnkiDroid 2.16.5 on Android 13 smartphone.
Nevertheless, I was able to workaround that blocking issue by doing the following:

1. Cleaned up AnkiDroid app stored data (perhaps/probably optional)
2. Uninstalled AnkiDroid 2.16.5 app
3. Via Neo Store (F-Droid-like app), I managed to install older AnkiDroid 2.16.3 **=> OK** (I was then able to: launch AnkiDroid, connect, sync and use app without any blocker)
4. Still via Neo Store, I managed to upgrade to AnkiDroid 2.16.4 **=> OK** (I was able to: launch AnkiDroid, then was asked to: "grant access to all files (special access)), then, once access granted, I was able to: connect, sync and use app without any blocker)
5. Still via Neo Store, I upgraded to AnkiDroid 2.16.5 **=> OK** (I was able to: launch AnkiDroid, connect, sync and use app without any blocker)

If I compare this situation to my initial launch of AnkiDroid (blocked by current issue), here is the main difference: First time, I was launching AnkiDroid (2.16.5) after a long period of inactivity (and multiple intermediate AnkiDroid updates), so last time I launched it, it was a version v2.x<2.16 (at that time, grant special access to all files was not yet required).

So from my own experience, and only based on it, here are my conclusions on that issue:

- It appears that 2.16.4 (upgrading a fresh 2.16.3 install) behaves correctly and request user to grant AnkiDroid app to "access all files".
- It appears that 2.16.5 (upgrading from old running 2.1x install) behaves unexpectedly and drives use to app properties/permissions, while it should drive it to: "special access" > "all files access" > AnkiDroid ( v ).

Hope this help.

Best regards,

**Comment by gregyski** [comment]
I experienced this as well.

Device: Pixel 7 Pro
Android version: 13 
Android security update: 2023-08-05
Ankidroid version: 2.16.5

I exhibited the same symptoms as ByteCodeRoad's video, additionally looking under "...->All Permissions" and not finding Storage there either. Deleting app data as per Ezkielnet and then logging back in and syncing solved it.

**Comment by snowtimeglass** [comment]
@Ezkielnet @gregyski 
Thank you for sharing information about the unexpected discovery. 

Just to be sure, after tapping the "Get started" in the initial screen, were you actually able to see the deck list screen? (And it was without being asked for any sort of permissions?)


(I don't have an Android 13 device so I can't try this way myself.)

**Comment by Ezkielnet** [comment]
> > when I hit Delete app data, the app opened.
> 
> > Deleting app data as per Ezkielnet and then logging back in and syncing solved it.
> 
> @Ezkielnet @gregyski Thank you for sharing information about the unexpected discovery.
> 
> Just to be sure, after tapping the "Get started" in the initial screen, were you actually able to see the deck list screen? (And it was without being asked for any sort of permissions?)
> 
> (I don't have an Android 13 device so I can't try this way myself.)

Yes, I can use the application perfectly. I can't be sure if it asks for permissions afterwards because I enabled them manually.

**Comment by snowtimeglass** [comment]
>Yes, I can use the application perfectly.

How nice. Thank you.

>I can't be sure if it asks for permissions afterwards because I enabled them manually.

Which permission items did you manually enable?
(Are those in the ByteCodeRoad's video?)

**Comment by Ezkielnet** [comment]
> Which permission items did you manually enable? (Are those in the ByteCodeRoad's video?)

When I had the error I enabled all of them, but I just saw that none of them are active. 

![WhatsApp Image 2023-09-30 at 14 47 36_c3c03829](https://github.com/ankidroid/Anki-Android/assets/23616105/310a9d14-18e8-4c5d-9378-e70fe5f0e02e)

**Comment by gregyski** [comment]
> > when I hit Delete app data, the app opened.
> 
> > Deleting app data as per Ezkielnet and then logging back in and syncing solved it.
> 
> @Ezkielnet @gregyski Thank you for sharing information about the unexpected discovery.
> 
> Just to be sure, after tapping the "Get started" in the initial screen, were you actually able to see the deck list screen? (And it was without being asked for any sort of permissions?)
> 
> (I don't have an Android 13 device so I can't try this way myself.)

If I"m understanding the question properly, before the "fix" I'd launch the app and be taken directly to permissions without ever seeing any of the Ankidroid interface. After the "fix", for all practical purposes I was starting with a fresh installation so I had to log in and sync to get my decks back from Ankiweb. I did not, at that time, get asked for any further permissions. Note that I had, when trying to fix the issue originally, manually enabled the "Read and write to the Ankidroid database" permission so that was set as well, so while that had no impact on the original problem, I don't know if it impacted the "fix". Hope that helps.

**Comment by snowtimeglass** [comment]
So, after the "Delete app data", the prior manual permissions may be no impact or annulled, and the app works properly without requiring permissions... I think I have a better understanding of the situation. Thank you for the feedbacks.

**Comment by srishti-R** [comment]
Hi, is anyone working on this? I would like to contribute.

**Comment by snowtimeglass** [comment]
No, I'm not. I welcome your contribution.

**Comment by snowtimeglass** [comment]
@srishti-R

 >If no one is assigned on a issue you can just comment that you are working on it and start working : )

_Originally posted by @criticalAY in https://github.com/ankidroid/Anki-Android/issues/14502#issuecomment-1746660974_

**Comment by srishti-R** [comment]
Strange that I am not facing any such issue in master branch.

**Comment by gregyski** [comment]
I believe it's possible that it might be related to the Android feature that disables permissions when an app hasn't been used in a while. I know I'd been slacking and am pretty sure that feature activated on my Ankidroid install. And it might even related to some recent Android permission changes making it rather an edge case if it's a combination of the two.

I have no evidence of that, but I know when I was researching the problem to try to fix it I encountered mentions of both situations being potential causes, so I thought I'd mention it.

**Comment by 7merlin** [comment]
Please fix or help find a workaround that doesn't involve deleting data. My AnkiWeb account expired so I'm concerned with loosing all the data if I clear it

**Comment by queue-tip** [comment]
> Please fix or help find a workaround that doesn't involve deleting data. My AnkiWeb account expired so I'm concerned with loosing all the data if I clear it

Either of these should work
- export your collection, including scheduling information and media, reinstall after deleting data, and then import the collection
- create an ankiweb account, sync to it, reinstall after deleting data, and sync again

**Comment by risshuu** [comment]
> I believe it's possible that it might be related to the Android feature that disables permissions when an app hasn't been used in a while. I know I'd been slacking and am pretty sure that feature activated on my Ankidroid install. And it might even related to some recent Android permission changes making it rather an edge case if it's a combination of the two.
> 
> I have no evidence of that, but I know when I was researching the problem to try to fix it I encountered mentions of both situations being potential causes, so I thought I'd mention it.

This would line up with what I am seeing.  This happened to me just recently.  I haven't used it in months.  I am in the same boat as 7merlin where I do not have the decks on ankiweb anymore because of the months that I have spent away from the app.  I would also like to be able to get this fixed without wiping my data.

queue-tip: I don't know how I am supposed to do your steps since I am getting this error and cannot get into the app.

**Comment by snowtimeglass** [comment]
>Please fix or help find a workaround that doesn't involve deleting data. My AnkiWeb account expired so I'm concerned with loosing all the data if I clear it

>I would also like to be able to get this fixed without wiping my data.

@7merlin @risshuu 
Long press the app icon > "App Info" > "Storage usage" > "Clear cache" doesn't work as another workaround?

**Comment by 7merlin** [comment]
No, it didn't help

**Comment by 7merlin** [comment]
[queue-tip](https://github.com/queue-tip), yes, same as [risshuu](https://github.com/risshuu) I can't follow the steps since the app is not opening. How can I export the collection if the app is not opening? How can I sync to ankiweb if the app is not opening?

**Comment by dae** [comment]
You could try clearing the app's cache (not data) to see if that helps.

https://www.avg.com/en/signal/how-to-clear-cache-on-android-phones#:~:text=a%20Samsung%20phone%3A-,Open%20Settings%20and%20select%20Apps.,Clear%20data%20and%20then%20OK.

Before trying that, if you don't have any other copies of your data, you may be able to extract the data by using a file browsing app that lets you read the files stored inside AnkiDroid's folder. I'm not an Android user, so perhaps someone else here can suggest a good one.

**Comment by jkexbx** [comment]
I'm also experiencing this on a oneplus 8T. It's definitely related to permissions that are removed due to inactivity.

First tried permissions, but there's nothing more I can give.

Then tried uninstalling and installing while keeping and then removing my data.

Finally, I left the google beta tester group, and reinstalled and that fixed it.

Installing 2.16.5 not from Beta fixed it.

Full video of process: https://youtu.be/IQJwKJLohoE

**Comment by 7merlin** [comment]
[dae](https://github.com/dae), I've tried clearing cache, it didn't help.

Where does AnkiDroid store files? Is it somewhere on /sdcard that's accessible via ADB?

**Comment by dae** [comment]
It should be accessible to ADB, but I think it's on the internal storage by default.

**Comment by david-allison** [comment]
I've been away for a while, may need some more time to process before I dive into this one

**Proposed resolution for advanced users:**

❓❗️ Could an affected user try the following ADB commands:

`adb shell pm grant com.ichi2.anki android.permission.WRITE_EXTERNAL_STORAGE`
`adb shell pm grant com.ichi2.anki android.permission.READ_EXTERNAL_STORAGE`

An affected user would have:
* Installed from the Play Store & not uninstalled
* OR 
    * installed AnkiDroid 'play' 2.15 
    * Upgraded to 2.16 
    * Had their phone remove permissions

If you're affected in another way, and the command works, please let me know. If you feel you're affected but don't know how to use adb, please ask

----

I'm going to summarise my understanding of this issue, so we can work forward to fix/inform users of a path to resolve this.

Please correct me if I'm incorrect

**Information for readers**

* AnkiDroid used to store data in `/AnkiDroid`. 
   * After an upgrade from 2.15 to 2.16, `/AnkiDroid` is still accessible
   * After uninstalling, `/AnkiDroid` is no longer accessible
* Installs from GitHub & Google Play are different
    * AnkiDroid 2.16 from the Play Store requires no storage permissions:  it cannot request `MANAGE_EXTERNAL_STORAGE
    * Otherwise, `MANAGE_EXTERNAL_STORAGE` allows us to 
* The storage permissions we care about is
  * AnkiDroid 2.15 - 'Files Permission'
     * This permission is no longer visible in the permissions screen when upgrading to 2.16, but remains granted (tested on Emulator)

**Red Herrings**
* 'Read and write to the AnkiDroid database' allows other apps to access the AnkiDroid API
* AnkiDroid 2.16.4 => AnkiDroid 2.16.5 had [effectively no functional changes](https://github.com/ankidroid/Anki-Android/compare/v2.16.4...v2.16.5)

**Issue Prerequisites**
* AnkiDroid has been updated from 2.15 to 2.16
*❓ Did the users with the issue perform the storage migration (Is AnkiDroid is storing files in the legacy location (`/AnkiDroid`)?)

**Trigger**
* AnkiDroid is not used in a while, and specific manufacturers will revoke unused permissions 
* ❓ What does this look like in terms of Permissions & `isExternalStorageLegacy`
   * My suspicion is that `WRITE_EXTERNAL_STRORAGE` is removed AND the app is in 2.16
* `Files Permission` is removed from the permissions screen once the app is updated to 2.16, so a user may not re-grant this
   * I don't **think** this is our issue, but we should audit [AndroidManifest.xml](https://github.com/ankidroid/Anki-Android/blob/main/AnkiDroid/src/main/AndroidManifest.xml)

----

Outcome of the request at the top of the post

* If the command works, we've diagnosed the issue, but I don't believe Android provides us any means to re-grant this permissions on the phone
* We can detect the situation and give advanced users instructions
* I suspect the user did **not** perform the storage migration
   * If they did, this is a bug we can fix
   * If they didn't, they can install a version of AnkiDroid which can request `MANAGE_EXTERNAL_STORAGE`

**Comment by eumel2** [comment]
Android 13
Installed via playstore and not used in a while though i did grant the permission from the settings to no avail
Version 2.16.5

I tried the adb commands and that fixed it for me @david-allison

**Comment by david-allison** [comment]
@eumel2 Fantastic! This feedback is vital as I can't reproduce the issue

Could you list your AnkiDroid Directory in `Settings - Advanced`?

**Comment by WhtUsrWrks** [comment]
Does anyone have a dummies' guide to running ADB? I've enabled usb debugging, connected to computer, installed ADB app from play store, and connected but the pc says a whole long screen of things like adb : The term 'adb' is not recognized as the name of a cmdlet, function, script file, or operable program. Check the
spelling of the name, or if a path was included, verify that the path is correct and try again.
At line:8 char:1
+ adb shell pm grant com.ichi2.anki android.permission.WRITE_EXTERNAL_S ...
+ ~~~
    + CategoryInfo          : ObjectNotFound: (adb:String) [], CommandNotFoundException
    + FullyQualifiedErrorId : CommandNotFoundException

and the app just shows that I am connected and I can see where I pasted the adb commands but hasn't done anything

**Comment by david-allison** [comment]
There is no ADB app.

**Get ADB working**
Once this step is completed, running `adb --version` on the command line should produce output

* Download and install the Platform Tools: https://developer.android.com/tools/releases/platform-tools
* See if `adb` works in the terminal/command line
* If it doesn't, open a command line in the `android_sdk/platform-tools/` folder which you installed 
   * Shift + right click in a folder in Windows and you should see the option to open a terminal/command prompt there
 
----
* Enable Developer Options on your Android
* Enable USB/Wireless Debugging
* Plug in your device to your computer via USB
* Enable your computer debug access to your phone (A prompt will appear on your Android)
* run `adb devices` and see if your device is listed

**Comment by david-allison** [comment]
The next stage in fixing this is determining when the bug takes place:

* API 33 ⚠️ `WRITE_EXTERNAL_STRORAGE` cannot be set in 21.6
* API 32 ✅ `WRITE_EXTERNAL_STRORAGE` can be set in 2.16

<details>

<summary>Screenshots</summary>

API 32 on 21.6:
<img width="439" alt="Screenshot 2023-10-24 at 16 32 29" src="https://github.com/ankidroid/Anki-Android/assets/62114487/552e00d3-9cf6-40ce-a154-33adcb8fda7c">

Permission can be revoked and reapplied

**API 33**

**2.15:**
<img width="416" alt="Screenshot 2023-10-24 at 16 35 00" src="https://github.com/ankidroid/Anki-Android/assets/62114487/f18f4c3a-d5d0-4f71-a0ee-e0d8011870cf">
<img width="418" alt="Screenshot 2023-10-24 at 16 35 32" src="https://github.com/ankidroid/Anki-Android/assets/62114487/bc350ea0-bd4f-4112-a296-97f57b66cca4">

**2.16:**
<img width="408" alt="Screenshot 2023-10-24 at 16 35 59" src="https://github.com/ankidroid/Anki-Android/assets/62114487/f278bd06-f6b3-4fbd-a68b-cba979c216ea">

</details>

**Comment by david-allison** [comment]
# Issue
* If a user is on API 33 [Android 13] and maybe higher
* And the user has not migrated storage
* Android can remove `WRITE_EXTERNAL_STRORAGE` if the app is inactive, with no means to reapply this permission via the Android UI
* The user can no longer access the `/AnkiDroid` folder
* 🪲 AnkiDroid requests the user applies permissions, but this is not possible


# User actions

## Advanced User
**An advanced user can fix this issue via ADB:** 

```
adb shell pm grant com.ichi2.anki android.permission.WRITE_EXTERNAL_STORAGE
adb shell pm grant com.ichi2.anki android.permission.READ_EXTERNAL_STORAGE
```

## Semi-Advanced User

Download and use the `full` version of AnkiDroid (cannot be distributed by Google Play) which requests `MANAGE_EXTERNAL_STORAGE` and can access `/AnkiDroid`

https://github.com/ankidroid/Anki-Android/releases/latest

## Any user
Reinstall AnkiDroid from the Play Store

* ⚠️ Due to Android OS policies, the user permanently loses access to `/AnkiDroid`
* This makes a new collection folder in `/Android/data/...`
* A user can sync via AnkiWeb
  * It is unlikely that the user is syncing, as they did not migrate storage
* A user can import a collection backup from the `/AnkiDroid` folder
* ⚠️ A user may or may not be able to re-import their media

# What should we do?

This is up for discussion

* We should inform the user of this issue, using a dialog similar to `DIALOG_STORAGE_UNAVAILABLE_AFTER_UNINSTALL`
* We should provide the same options as the dialog above, with an additional option for advanced users to fix the issue via ADB

**Comment by eumel2** [comment]
The folder is: /storage/emulated/0/AnkiDroid
Starting it just now i also got a message to migrate the storage somewhere which I postponed for now

**Comment by snowtimeglass** [comment]
@david-allison Thank you for putting the issue in perspective.

What do you think about the `App info - Storage usage - Manage space - Delete app data` approach, which @Ezkielnet and @gregyski have reported. If I understand correctly, it won't delete the collection, and it will enable to open the app and access the `/AnkiDroid` folder again, without uninstalling the Play Store app.
![image](https://github.com/ankidroid/Anki-Android/assets/10436072/65717a76-ab61-4861-9ca1-def8b3add5ef)
(Screenshot of the screen of Android 9, not 13)

**Comment by david-allison** [comment]
@snowtimeglass  I don't /think/ that would work: 

https://github.com/ankidroid/Anki-Android/blob/b44dcf49d5a410751ead0d29b7b3c90ccbfed965/AnkiDroid/src/main/java/com/ichi2/anki/InitialActivity.kt#L287-L295

I believe the issue is that `Environment.isExternalStorageLegacy()` returns true BUT the user can't provide permissions to access the legacy storage. Therefore the new default directory would be unsafe

Currently, a user would need to uninstall to remove legacy storage access, 

----

`DIALOG_STORAGE_UNAVAILABLE_AFTER_UNINSTALL ` offers "Create a new collection" which is a less destructive version of 'Delete app data', and would fix the issue if it sets `collectionPath` to the app-private directory


![Screenshot 2023-03-04 at 03 07 40](https://user-images.githubusercontent.com/62114487/222872922-3e1cfee1-bcbc-4bb7-a0c7-324a89fb1108.png)

**Comment by snowtimeglass** [comment]
@david-allison 

>I don't /think/ that would work:

>Therefore the new default directory would be unsafe

In this case, what do "not work" and "unsafe" mean specifically? You mean that the approach may make the legacy directory inaccessible?

**Comment by david-allison** [comment]
I believe that clearing app data would mean that:

* Preferences are cleared: `collectionPath` being the main impact
* `collectionPath` is set to the default, which I believe would be `/AnkiDroid` (as `Environment.isExternalStorageLegacy` is `true`)
* This is still unusable for the user, as the user still does not have permission to access it

> What do[es] ... "unsafe" mean specifically?

I misspoke: I edited the comment to replace unsafe with unusable

**Comment by gregyski** [comment]
> @david-allison What do you think about the `App info - Storage usage - Manage space - Delete app data` approach, which @Ezkielnet and @gregyski have [reported](https://github.com/ankidroid/Anki-Android/issues/14423#issuecomment-1726195968). If I understand correctly, it won't delete the collection, and it will enable to open the app and access the `/AnkiDroid` folder again, without uninstalling the Play Store app.

If I remember correctly, Delete app data did delete my decks. But since I still had them on Ankiweb I was able to download them again.

**Comment by snowtimeglass** [comment]
> * `collectionPath` is set to the default, which I believe would be `/AnkiDroid` (as `Environment.isExternalStorageLegacy` is `true`)

I suppose so, too.

> * This is still unusable for the user, as the user still does not have permission to access it

Can't the [video](https://github.com/ankidroid/Anki-Android/issues/14423#issuecomment-1726195968) of @Ezkielnet above (and subsequent comments) disprove the prediction?

**Comment by snowtimeglass** [comment]
>If I remember correctly, Delete app data did delete my decks. 

Although the conditions are different, the `Delete app data` doesn't delete the decks (collection) on my device (Android 9).

**Comment by snowtimeglass** [comment]
> so wiping app data would also work to allow a user to reuse the app without uninstalling (same outcome though: `/AnkiDroid` is inaccessible, so it appears that decks have been deleted).

> the Anki collection should still be intact in `~/AnkiDroid` after deleting app data, but since AnkiDroid can no longer access this folder, it appears that a new collection is created from scratch.

Excuse me, the following two thoughts are still not consistent in my mind. 

- After executing `Delete App data`, AnkiDroid can no longer access `/AnkiDroid` folder

 - After executing `Delete App data`, @Ezkielnet and @gregyski can now use the app as before (aside from whether the previous collection is usable or not)"

 A new collection is supposed to be created in a new `/AnkiDroid` folder and AnkiDroid can access it?

**Comment by david-allison** [comment]
The new collection is created in `/Android/data/com.ichi2.anki/files/AnkiDroid`

**Comment by snowtimeglass** [comment]
It is created automatically without this dialog?
![image](https://github.com/ankidroid/Anki-Android/assets/10436072/e083d7aa-185c-465f-bed3-a085a027c3c1)

**Comment by david-allison** [comment]
Yes, "Create a new collection" will call through to:

https://github.com/ankidroid/Anki-Android/blob/603bd59a20c15037d0eea1297e22178dce6b76a0/AnkiDroid/src/main/java/com/ichi2/anki/CollectionHelper.kt#L502-L512


`/Android/data/com.ichi2.anki/files/AnkiDroid` is the new default location of the AnkiDroid folder

https://github.com/ankidroid/Anki-Android/blob/603bd59a20c15037d0eea1297e22178dce6b76a0/AnkiDroid/src/main/java/com/ichi2/anki/CollectionHelper.kt#L334-L441

**Comment by snowtimeglass** [comment]
Ah, I think I understand now. Thank you so much for your time and detailed explanation.

**Comment by chmnchiang** [comment]
Just want to report that "Delete app data" deleted all my decks (and I couldn't sync to web because I can't open the app 🥹)... So probably not a good general solution....

**Comment by david-allison** [comment]
> Just want to report that "Delete app data" deleted all my decks (and I couldn't sync to web because I can't open the app 🥹)... So probably not a good general solution.... 

Does your phone have an `AnkiDroid` folder on the root of the device storage?

That should contain your data.

Installing the `full` APK from GitHub should allow you to access it again

https://github.com/ankidroid/anki-android/releases/latest

**Comment by chmnchiang** [comment]
That works. Thanks

**Comment by dimdenGD** [comment]
I got this bug and here's some info, maybe will be useful to anyone (checked on [2.17alpha8](https://github.com/ankidroid/Anki-Android/releases/tag/v2.17alpha8) Android 13): 
- When opening AnkiDroid for first time after installation there's 2 buttons to either get started or sync, if I press sync, I get `com.ichi2.anki.exception.StorageAccessException: Failed to create AnkiDroid directory /storage/emulated/0/AnkiDroid` error.
- Going to app permissions for some reason doesn't show storage permission
- Pressing get started button on that menu prompts me to give storage permission and then I can login to sync from settings and it works fine

**Comment by david-allison** [comment]
@dimdenGD that seems to be an instance of the following, which will be possible to resolve:

* #14650

**Comment by ben-albrecht** [comment]
I had this issue recently. Uninstalling the app without keeping local data, then reinstalling worked for me. I selected "create a new collection"  when opening it after reinstallation. 

However, this will wipe your decks if you don't have your collection synced to AnkiWeb.

**Comment by hanpingchinese** [comment]
Out of interest, as an alternative to installing the universal APK from github, couldn't you suggest the user copies their AnkiDroid folder to `/sdcard/Android/data/com.ichi2.anki/files`?

**Comment by david-allison** [comment]
Android file handling has been painful for the past few releases and the user experience would (likely) be much worse: 

* Most file managers can't copy `/AnkiDroid` after importing the standard medschool deck due to the size and number of files in `collection.media`
* Most on-device file managers don't have access to `/Android`, a user needs a PC, and a number of people are mobile-only
* USB transfers have historically also been awful (untested in newer Android versions)
* `/AnkiDroid` -> `/Android/data` copies across mount points, so `rename(2)` won't work

**Comment by soshial** [comment]
Out of interest, why doesn't AnkiDroid simply store decks in its own folder, for which there would be no permissions required. Asking for generic STORAGE permission just to store and manage own files is a bad approach for an app.

**Comment by mikehardy** [comment]
@soshial you first assume that AnkiDroid doesn't store decks in it's own folder.

That is an incorrect assumption for current AnkiDroid, so the rest of your comment is invalid.

For older versions (which were released way back with Android 1!) there was no "own folder". So AnkiDroid used `/sdcard/AnkiDroid`.

Now, what do you do as an app when you have users from older versions of Android and AnkiDroid that still have their precious data in `/sdcard/AnkiDroid` but your new version uses it's own folder? Why you migrate the data of course! 

Which was a huge programming effort. But it's done now.

However, to do that migration you still need access to `/sdcard/AnkiDroid` right? Now we're all caught up :-)

**Comment by TheMadSword** [comment]
Same issue @ Pixel 4a, Android 13. When installing latest 2.17alpha14, I then get this screen : 

<img width="353" alt="image" src="https://github.com/ankidroid/Anki-Android/assets/2840135/e8c252aa-2aab-41be-9027-35eaa7a56bab">

What is weird : 
1- I don't recall ever having that specific permission screen before on my phone.
2- Whenever the toggle is used, dragging or clicking, it directly get back to "off". It cannot be activated.
3- All permissions I can give on my phone to AnkiDroid are given. The closest seems to be it is "Read and write to the AnkiDroid database". Others are Microphone and notifications...

Thanks !

**Comment by soshial** [comment]
That's great news that AnkiDroid will use internal folder for deck files @mikehardy. Thanks for the explanation.

Although v2.16 cannot be launched for me at all (I couldn't install latest 2.17alpha versions because of errors). When I open the app on my Android 14 it redirects me to the permissions screen and demands to allow storage permission, which is not in the list. @mikehardy

**Comment by mikehardy** [comment]
Use the full/universal build from our releases page. You say "errors" but don't specifically say what error. We're not aware of any thing that would stop am install of the alpha so that should work. Be more specific and we can probably help

**Comment by jgarvin** [comment]
Can confirm this still exists. I got the notice today it had been too long since login and my decks could get deleted, so I opened Ankidroid to try to get it to sync. It asks for storage permission and opens android app settings for Ankidroid. However the only permission listed is permission to "read and write to and from the Ankidroid database", which I granted, but the message still appears. Under "Not allowed" permissions it says "No permissions denied."

**Comment by mikehardy** [comment]
@jgarvin how did it go following this direction?

> Use the full/universal build from our releases page

...hopefully that worked

**Comment by soshial** [comment]
@mikehardy I wonder how this universal build differs from the one in the Google Play? Why cannot this build be published on Google Play then, if I may ask?

**Comment by david-allison** [comment]
> @mikehardy I wonder how this universal build differs from the one in the Google Play? Why cannot this build be published on Google Play then, if I may ask?

Google Play won't allow it: https://support.google.com/googleplay/android-developer/answer/10467955?hl=en

**Comment by alexanderadam** [comment]
> Google Play won't allow it, as they don't want apps to have access to the filesystem. https://support.google.com/googleplay/android-developer/answer/10467955?hl=en

And why is [the F-Droid release](https://f-droid.org/en/packages/com.ichi2.anki/) buggy as well then? :thinking:

**Comment by david-allison** [comment]
> > Google Play won't allow it, as they don't want apps to have access to the filesystem. https://support.google.com/googleplay/android-developer/answer/10467955?hl=en
> 
> And why is [the F-Droid release](https://f-droid.org/en/packages/com.ichi2.anki/) buggy as well then? 🤔

This occurs in the F-Droid release? Could you post a separate issue

**Comment by david-allison** [comment]
They either use `DocumentFile` + requesting access to a folder, which is insufficient for database access, or they use [`MediaStore`](https://developer.android.com/training/data-storage/shared/media#media_store), which is also insufficient for database access

**Comment by alexanderadam** [comment]
> This occurs in the F-Droid release? Could you post a separate issue

[done](https://github.com/ankidroid/Anki-Android/issues/15463) :+1:

**Comment by brishtibheja** [comment]
Weird. It is working fine with my Xiaomi device. I just simply used "set up my device" from Google and shared the app from my old phone using ShareMe.
I was using arm64 ver. from here. It was the second beta version. Maybe that's why it didn't asked me for permissions. People who are having problems may try syncing settings from a old device using their Google account.

**Comment by david-allison** [comment]
Partial patch if anyone wants to take this to completion. The outcome is awful, but our hands are tied. The main solace is that there's no data deletion, only revocation of AnkiDroid's access to the data.

TL;DR: If a user is on Tiramissu (Android 13/SDK 33) or above, Android can revoke `WRITE_EXTERNAL_STORAGE`/`READ_EXTERNAL_STORAGE` and the only way to reapply this is `adb`

Options:
* install from GitHub/F-Droid
* use `adb` to re-grant the permissions
* reinstall (moving to the app private folder) and
  * resync
  * restore from a backup in `/AnkiDroid` (manually copying media - which is problematic)



<details>

<summary>Patch</summary>

```patch
Subject: [PATCH] AndroidPermanentlyRevokedPermissionsDialog.kt
---
Index: AnkiDroid/src/main/res/values/01-core.xml
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
diff --git a/AnkiDroid/src/main/res/values/01-core.xml b/AnkiDroid/src/main/res/values/01-core.xml
--- a/AnkiDroid/src/main/res/values/01-core.xml	
+++ b/AnkiDroid/src/main/res/values/01-core.xml	
@@ -274,6 +274,7 @@
 
     <string name="directory_inaccessible_after_uninstall" comment="Dialog title if AnkiDroid can't access the collection once the app is installed">Inaccessible collection</string>
     <string name="directory_inaccessible_after_uninstall_summary" comment="the parameter is the path to the AnkiDroid folder. Typically /storage/emulated/0/AnkiDroid">We are unable to access your collection after AnkiDroid is uninstalled due to a change in Play Store Policy\n\nYour data is safe and can be restored. It is located at\n%s\n\nSelect an option below to restore:</string>
+    <string name="directory_revoked_after_inactivity" comment="the parameter is the path to the AnkiDroid folder. Typically /storage/emulated/0/AnkiDroid">We are unable to access your collection after AnkiDroid is uninstalled due to a change in Play Store Policy\n\nYour data is safe and can be restored. It is located at\n%s\n\nSelect an option below to restore:</string>
     <string name="restore_data_from_ankiweb">Restore from AnkiWeb (recommended)</string>
     <string name="install_non_play_store_ankidroid_recommended">Restore folder access (recommended)</string>
     <string name="install_non_play_store_ankidroid">Restore folder access (advanced)</string>
Index: AnkiDroid/src/main/java/com/ichi2/anki/ui/windows/permissions/AndroidPermanentlyRevokedPermissionsDialog.kt
===================================================================
diff --git a/AnkiDroid/src/main/java/com/ichi2/anki/ui/windows/permissions/AndroidPermanentlyRevokedPermissionsDialog.kt b/AnkiDroid/src/main/java/com/ichi2/anki/ui/windows/permissions/AndroidPermanentlyRevokedPermissionsDialog.kt
new file mode 100644
--- /dev/null	
+++ b/AnkiDroid/src/main/java/com/ichi2/anki/ui/windows/permissions/AndroidPermanentlyRevokedPermissionsDialog.kt	
@@ -0,0 +1,72 @@
+/*
+ *  Copyright (c) 2024 David Allison <davidallisongithub@gmail.com>
+ *
+ *  This program is free software; you can redistribute it and/or modify it under
+ *  the terms of the GNU General Public License as published by the Free Software
+ *  Foundation; either version 3 of the License, or (at your option) any later
+ *  version.
+ *
+ *  This program is distributed in the hope that it will be useful, but WITHOUT ANY
+ *  WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS FOR A
+ *  PARTICULAR PURPOSE. See the GNU General Public License for more details.
+ *
+ *  You should have received a copy of the GNU General Public License along with
+ *  this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+
+package com.ichi2.anki.ui.windows.permissions
+
+import android.annotation.SuppressLint
+import androidx.annotation.StringRes
+import com.afollestad.materialdialogs.MaterialDialog
+import com.afollestad.materialdialogs.list.listItems
+import com.ichi2.anki.AnkiActivity
+import com.ichi2.anki.R
+import com.ichi2.anki.dialogs.DatabaseErrorDialog
+import com.ichi2.anki.dialogs.DatabaseErrorDialog.*
+
+object AndroidPermanentlyRevokedPermissionsDialog {
+    @SuppressLint("CheckResult")
+    fun show(context: AnkiActivity) {
+        val listItemData = StoragePermanentlyRevokedOptions.createList()
+
+        val info = """Android 13 may permanently removes access to your AnkiDroid collection if 
+            |AnkiDroid is installed from Google Play
+            |
+            |Your data is safe, but inaccessible
+            |
+            |
+        """.trimMargin()
+        MaterialDialog(context).show {
+            message(res = R.string.directory_revoked_after_inactivity)
+            listItems(items = listItemData.map { context.getString(it.stringRes) }, waitForPositiveButton = false) { _: MaterialDialog, index: Int, _: CharSequence ->
+                val listItem = listItemData[index]
+                listItem.onClick(context)
+                if (listItem.dismissesDialog) {
+                    this.dismiss()
+                }
+            }
+            noAutoDismiss()
+            cancelable(false)
+        }
+    }
+
+    /**
+     * List items, copied from [DatabaseErrorDialog.UninstallListItem]
+     */
+    private class StoragePermanentlyRevokedOptions(@StringRes val stringRes: Int, val dismissesDialog: Boolean, val onClick: (AnkiActivity) -> Unit) {
+        companion object {
+            fun createList(): List<StoragePermanentlyRevokedOptions> {
+                return UninstallListItem.createList()
+                    .filter { it != UninstallListItem.RESTORE_FROM_ANKIWEB }
+                    .map { listItem ->
+                        StoragePermanentlyRevokedOptions(
+                            stringRes = listItem.stringRes,
+                            dismissesDialog = listItem.dismissesDialog,
+                            onClick = listItem.onClick
+                        )
+                    }
+            }
+        }
+    }
+}
\ No newline at end of file
Index: AnkiDroid/src/main/java/com/ichi2/anki/dialogs/DatabaseErrorDialog.kt
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
diff --git a/AnkiDroid/src/main/java/com/ichi2/anki/dialogs/DatabaseErrorDialog.kt b/AnkiDroid/src/main/java/com/ichi2/anki/dialogs/DatabaseErrorDialog.kt
--- a/AnkiDroid/src/main/java/com/ichi2/anki/dialogs/DatabaseErrorDialog.kt	
+++ b/AnkiDroid/src/main/java/com/ichi2/anki/dialogs/DatabaseErrorDialog.kt	
@@ -364,7 +364,7 @@
     }
 
     /** List items for [DIALOG_STORAGE_UNAVAILABLE_AFTER_UNINSTALL] */
-    private enum class UninstallListItem(@StringRes val stringRes: Int, val dismissesDialog: Boolean, val onClick: (DeckPicker) -> Unit) {
+    enum class UninstallListItem(@StringRes val stringRes: Int, val dismissesDialog: Boolean, val onClick: (AnkiActivity) -> Unit) {
 
         RESTORE_FROM_ANKIWEB(
             R.string.restore_data_from_ankiweb,
@@ -392,7 +392,8 @@
         RESTORE_FROM_BACKUP(
             R.string.restore_data_from_backup,
             dismissesDialog = false,
-            { deckPicker ->
+            { activity ->
+                val deckPicker = activity as DeckPicker
                 Timber.i("Restoring from colpkg")
                 val newAnkiDroidDirectory = CollectionHelper.getDefaultAnkiDroidDirectory(deckPicker)
                 deckPicker.importColpkgListener = DatabaseRestorationListener(deckPicker, newAnkiDroidDirectory)
@@ -429,7 +430,7 @@
 
         companion object {
             /** A dialog which creates a new collection in an unsafe location */
-            fun displayResetToNewDirectoryDialog(context: DeckPicker) {
+            fun displayResetToNewDirectoryDialog(context: AnkiActivity) {
                 AlertDialog.Builder(context).show {
                     title(R.string.backup_new_collection)
                     setIcon(R.drawable.ic_warning)
Index: AnkiDroid/src/main/java/com/ichi2/anki/ui/windows/permissions/PermissionsUntil29Fragment.kt
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
diff --git a/AnkiDroid/src/main/java/com/ichi2/anki/ui/windows/permissions/PermissionsUntil29Fragment.kt b/AnkiDroid/src/main/java/com/ichi2/anki/ui/windows/permissions/PermissionsUntil29Fragment.kt
--- a/AnkiDroid/src/main/java/com/ichi2/anki/ui/windows/permissions/PermissionsUntil29Fragment.kt	
+++ b/AnkiDroid/src/main/java/com/ichi2/anki/ui/windows/permissions/PermissionsUntil29Fragment.kt	
@@ -15,9 +15,11 @@
  */
 package com.ichi2.anki.ui.windows.permissions
 
+import android.os.Build
 import android.os.Bundle
 import android.view.View
 import androidx.activity.result.contract.ActivityResultContracts
+import com.ichi2.anki.AnkiActivity
 import com.ichi2.anki.R
 import com.ichi2.utils.Permissions
 import com.ichi2.utils.hasAnyOfPermissionsBeenDenied
@@ -44,5 +46,13 @@
                 showToastAndOpenAppSettingsScreen(R.string.startup_no_storage_permission)
             }
         }
+
+        if (!userCanGrantWriteExternalStorage()) {
+            AndroidPermanentlyRevokedPermissionsDialog.show(requireActivity() as AnkiActivity)
+        }
     }
+
+    // On SDK 33 (TIRAMISU), `WRITE_EXTERNAL_STORAGE` cannot be set after AnkiDroid 2.15
+    // https://github.com/ankidroid/Anki-Android/issues/14423#issuecomment-1777504376
+    private fun userCanGrantWriteExternalStorage() = Build.VERSION.SDK_INT < Build.VERSION_CODES.TIRAMISU
 }
```

</details>

**Comment by alexanderadam** [comment]
> They either use `DocumentFile` + requesting access to a folder, which is insufficient for database access, or they use [`MediaStore`](https://developer.android.com/training/data-storage/shared/media#media_store), which is also insufficient for database access

Would you mind elaborating why this is insufficient?

Usually apps allow syncing and/or exporting for keeping data over installations.
Can't Anki do the same?

Syncing seem to be supported already so I understand once there's an import/export possible, the deprecated permission isn't needed Anyway. Or am I missing something here?

What exactly is the advantage of this permission that makes it necessary.
Which feature wouldn't work any more once the permission is removed?
What exactly is working in Anki that wouldn't work anyway?

**Comment by david-allison** [comment]
A `DocumentFile` is an abstraction which can either be a file on the local system, or from a file provider such as Google Drive.

On the local file system, this can be between 20 and 100x slower for basic operations. Listing files can be significantly slower than this (if it works at all). These files do not support random access and can't be passed into any of the standard `File`-based APIs

On the network, the situation is worse. It depends on external providers, and there are known bugs where folder structures are not represented properly and the listing of files is incomplete

https://commonsware.com/blog/2019/12/01/scoped-storage-stories-problems-saf.html

In my opinion, it is not a safe API which we can use to copy an AnkiDroid collection & media. Especially as a failed transfer would permanently delete user data.

----

On the Play Store, we only support the removed storage permission if a user is using an older version of Android or if a user has upgraded from 2.15 to 2.16. A fresh install from the Play Store does not use any storage permissions on a modern version of Android. 

This issue occurs because a user did not migrate their storage to the new location when prompted, and Android has revoked their permission to the legacy location.

When AnkiDroid had access to the legacy location, it produced backups which can be restored from, BUT: these backups do not include media (we have a huge cohort of users with gigabytes of media)

**Comment by alexanderadam** [comment]
Thank you so much for your explanations!

> In my opinion, it is not a safe API which we can use to copy an AnkiDroid collection & media. Especially as a failed transfer would permanently delete user data.

So how come that users don't experience any deleted data with other apps?
For instance apps like [Round Sync](https://github.com/newhinton/Round-Sync?tab=readme-ov-file#round-sync---rclone-for-android) do heavy I/O and I never experienced any deleted data.

And what is the long-term plan now?
Keeping AnkiDroid builds completely unusable instead of risking to use the newer API?

**Comment by david-allison** [comment]
Round Sync uses `MANAGE_EXTERNAL_STORAGE`, which we can't use on Google Play:

https://github.com/newhinton/Round-Sync/blob/master/app/src/main/AndroidManifest.xml

----

The long term plan is above: https://github.com/ankidroid/Anki-Android/issues/14423#issuecomment-1963007547

**Comment by mikehardy** [comment]
I think that solution is fine in general (https://github.com/ankidroid/Anki-Android/issues/14423#issuecomment-1963007547)

We are already nagging people about doing the migration and syncing, I'm not sure what more we can do?

The string in your proposed patch could perhaps be less specific about no longer able to access after uninstall and more about "Your collection is in the shared storage area of your device, but we no longer have access to it and Google no longer permits us to request access to it." ...then the other stuff. It could be the system revoked permission, it could  be they uninstalled/reinstalled, it could be the user themselves revoked permission... we're not sure why but we don't have permission so the string might be genericized a bit to reflect that

**Comment by david-allison** [comment]
My patch definitely needs some TLC, I won't have time to bring it through

* move the warning to when the user presses the button asking for permission
* string fixes
* (possibly) go through and handle 'restore backup'

**Comment by alexanderadam** [comment]
> Round Sync uses `MANAGE_EXTERNAL_STORAGE`, which we can't use on Google Play:
> 
> https://github.com/newhinton/Round-Sync/blob/master/app/src/main/AndroidManifest.xml

I think it's also capable of [using SAF](https://github.com/newhinton/Round-Sync/blob/2b1bfb5eb1aa20d59db2103b35312ff1edc4251e/safdav/src/main/java/io/github/x0b/safdav/SafAccessProvider.java#L16).
[It can't do random seek on large file (300 MB) in this example](https://github.com/newhinton/Round-Sync/issues/28) but I don't know whether this is relevant to Anki.

> The long term plan is above: [#14423 (comment)](https://github.com/ankidroid/Anki-Android/issues/14423#issuecomment-1963007547)

Got it :+1: 

btw I really appreciate your communication to clarify things. :pray:

**Comment by TheAdityaGupta** [comment]
May I know what is the current status of this write external storage issue?
What is done? 
What Can't be done?
What needs to be done? etc.
